### PR TITLE
Add do notation for Array

### PR DESCRIPTION
**Pull request by KhraksMamtsov**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Do notation (pipe version) for Array module.

## Related

- Related Issue #
- Closes #

### CODE REVIEW

File: packages/effect/src/Array.ts [hunk 1]
```diff
@@ -2114,3 +2114,82 @@ export const cartesian: {
...
+// -------------------------------------------------------------------------------------
+
+/**
+ * @since 2.4.0
+ * @category do notation
+ */
+export const Do: Array<{}> = of({})
+
+/**
+ * Binds an effectful value in a `do` scope
```
Comment (datner) [inline packages/effect/src/Array.ts:2129]:
```suggestion
 * Binds an array value in a `do` scope
```

**Review by datner — COMMENTED** [review]
This looks fine, but I must know:
_why_? What can this be used for? 🤔 

Also please add more jsdocs, preferably with example 🙏🏻

**Comment by mikearnaldi** [comment]
> This looks fine, but I must know: _why_? What can this be used for? 🤔
> 
> Also please add more jsdocs, preferably with example 🙏🏻

Do in array is as important as do in stream it's used to build up on products basically (zipWith on steroids)

**Comment by KhraksMamtsov** [comment]
> This looks fine, but I must know: _why_? What can this be used for? 🤔
> 
> Also please add more jsdocs, preferably with example 🙏🏻

I use this to generate independent props for stories in storybook
```ts
const btnProps = pipe(
    RA.Do,
    RA.bind("size", () => ["small", "medium", "large"] as const),
    RA.bind("theme", () => ["dark", "light", "contrast-light", "contrast-dark"] as const),
    RA.bind("disabled", () => [false, true]),
    RA.bind("loading", () => [false, true]),
    RA.bind("leftIcon", () => [null, "arrow-left", "disk"]),
    RA.bind("rightIcon", () => [null, "arrow-right", "save"]),
    RA.bind("text", () => [lorem(1), lorem(3), lorem(10)])
)
```

**Comment by gcanti** [comment]
can be used to simulate "array comprehension"

> Array comprehension is a technique in programming that allows you to create new arrays by iterating over existing ones and applying specific **conditions** or **transformations** to the elements. It's like assembling a new collection from pieces of other collections based on certain rules.

Haskell

```haskell
[(x,y) | x <- [1,3,5], y <- [2,4,6], x < y]
-- Output: [(1,2),(1,4),(1,6),(3,4),(3,6),(5,6)]
```

Effect

```ts
import { Array as Arr, pipe } from "effect"

const result = pipe(
  Arr.Do,
  Arr.bind("x", () => [1, 3, 5]),
  Arr.bind("y", () => [2, 4, 6]),
  Arr.filter(({ x, y }) => x < y), // condition
  Arr.map(({ x, y }) => [x, y] as const) // transformation
)

console.log(result)
/*
Output:
[ [ 1, 2 ], [ 1, 4 ], [ 1, 6 ], [ 3, 4 ], [ 3, 6 ], [ 5, 6 ] ]
*/
```

We could use this short explanation / example (without the Haskell bit) in the docs

@KhraksMamtsov I've merged this PR that makes it easier to define the do notation for Array https://github.com/Effect-TS/effect/pull/2681

**Comment by KhraksMamtsov** [comment]
@gcanti If you don't mind, I'll add 
```ts
// doNotation.ts
export type EmptyScope<F extends TypeLambda> = Kind<F, never, never, never, {}>
// module
export const Do: doNotation.EmptyScope<ModuleTypeLambda> = ...
```

**Comment by gcanti** [comment]
I think our convention is not to export anything public from the `internal` folder

**Comment by datner** [comment]
> Do in array is as important as do in stream it's used to build up on products basically (zipWith on steroids)

I conceptualize this well 🤔 

but the size doesn't explode?

```ts
const btnProps = pipe(
    RA.Do,
    RA.bind("size", () => ["small", "medium", "large"] as const),
    RA.bind("theme", () => ["dark", "light", "contrast-light", "contrast-dark"] as const),
    RA.bind("disabled", () => [false, true]),
    RA.bind("loading", () => [false, true]),
    RA.bind("leftIcon", () => [null, "arrow-left", "disk"]),
    RA.bind("rightIcon", () => [null, "arrow-right", "save"]),
    RA.bind("text", () => [lorem(1), lorem(3), lorem(10)])
)
```
thats an array of 1,296 members, no?

This is not to say not to add this, I'm purely interested in the concept and utility

File: packages/effect/src/Array.ts [hunk 2]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
+ *       result = [];
+ * for(let i = 0; i < x.length; i++) {
+ *   for(let j = 0; j < y.length; j++) {
+ *     const _x = x[i], _y = y[j];
+ *     if(_x < _y) result.push([_x, _y] as const)
+ *   }
+ * }
+ *
+ * @category do notation
+ * @since 2.0.0
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2164]:
```suggestion
 * @since 3.2.0
```

File: packages/effect/src/Array.ts [hunk 3]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2209]:
```suggestion
 * @since 3.2.0
```

File: packages/effect/src/Array.ts [hunk 4]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2266]:
```suggestion
 * @since 3.2.0
```

File: packages/effect/src/Array.ts [hunk 5]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
+   *       result = [];
+   * for(let i = 0; i < x.length; i++) {
+   *   for(let j = 0; j < y.length; j++) {
+   *     const _x = x[i], _y = y[j];
+   *     if(_x < _y) result.push([_x, _y] as const)
+   *   }
+   * }
+   *
+   * @category do notation
+   * @since 2.0.0
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2327]:
```suggestion
   * @since 3.2.0
```

File: packages/effect/src/internal/doNotation.ts [hunk 6]
```diff
@@ -16,6 +16,8 @@ type FlatMap<F extends TypeLambda> = {
   ): Kind<F, R1 & R2, O1 | O2, E1 | E2, B>
 }
 
+export type EmptyScope<F extends TypeLambda> = Kind<F, never, never, never, {}>
```
Comment (gcanti) [inline packages/effect/src/internal/doNotation.ts:19]:
```suggestion
/** @internal */
export type EmptyScope<F extends TypeLambda> = Kind<F, unknown, never, never, {}>
```

Comment (gcanti) [reply packages/effect/src/internal/doNotation.ts:19]: in the end this type seems superfluous to me

File: packages/effect/src/Array.ts [hunk 7]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
+ *   for(let j = 0; j < y.length; j++) {
+ *     const _x = x[i], _y = y[j];
+ *     if(_x < _y) result.push([_x, _y] as const)
+ *   }
+ * }
+ *
+ * @category do notation
+ * @since 2.0.0
+ */
+export const Do: Array<{}> = of({})
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2166]:
```suggestion
export const Do: ReadonlyArray<{}> = of({})
```

**Review by gcanti — CHANGES_REQUESTED** [review]

File: packages/effect/src/internal/doNotation.ts (hunk 6 above)
Comment (KhraksMamtsov) [reply packages/effect/src/internal/doNotation.ts]: @gcanti, I've gotten rid of it.

**Review by gcanti — APPROVED** [review]

File: packages/effect/src/Array.ts [hunk 8]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
+ *
+ * @category do notation
+ * @since 3.2.0
+ */
+export const bind: {
+  <A extends object, N extends string, B>(
+    tag: Exclude<N, keyof A>,
+    f: (a: A) => ReadonlyArray<B>
+  ): (
+    self: Array<A>
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2216]:
```suggestion
    self: ReadonlyArray<A>
```

File: packages/effect/src/Array.ts [hunk 9]
```diff
@@ -2114,3 +2115,216 @@ export const cartesian: {
...
+  ) => Array<{ [K in N | keyof A]: K extends keyof A ? A[K] : B }>
+  <A extends object, N extends string, B>(
+    self: Array<A>,
```
Comment (gcanti) [inline packages/effect/src/Array.ts:2219]:
```suggestion
    self: ReadonlyArray<A>,
```

**Review by gcanti — CHANGES_REQUESTED** [review]

**Review by gcanti — APPROVED** [review]

**Pull request merged**